### PR TITLE
Fix status for missing relations

### LINKED ISSUE #144: cos-proxy goes to `Blocked` if it is configured to be able to send data downstream but does not have any sources of that data
### Bug Description

As reported by Nishant, he has cos-proxy in a blocked state:

```
(truncated)

App                   Version  Status   Scale  Charm                       Channel        Rev  Exposed  Message
cos-proxy             n/a      blocked      1  cos-proxy                   latest/edge     81  no       Missing one of (Grafana|dashboard|grafana-agent) relation(s)
grafana-agent                  active      13  grafana-agent               latest/stable   65  no       
landscape-client               active       9  landscape-client            latest/stable   69  no       Client registered!
ntp                   4.2      active      12  ntp                         latest/stable   50  no       chrony: Ready
ubuntu-advantage               active      13  ubuntu-advantage            latest/stable   79  no       Attached (esm-apps,esm-infra,livepatch)

Unit                     Workload  Agent  Machine  Public address  Ports           Message
cos-proxy/0*             blocked   idle   12       10.27.64.20                     Missing one of (Grafana|dashboard|grafana-agent) relation(s)
  grafana-agent/10       active    idle            10.27.64.20                     
  landscape-client/8     active    idle            10.27.64.20                     Client registered!
  ntp/13                 active    idle            10.27.64.20     123/udp         chrony: Ready
  ubuntu-advantage/8     active    idle            10.27.64.20                     Attached (esm-apps,esm-infra,livepatch)

Integration provider                                      Requirer                                                     Interface                Type         Message
cos-proxy:cos-agent                                       grafana-agent:cos-agent                                      cos_agent                subordinate  
cos-proxy:downstream-grafana-dashboard                    cos-grafana-dashboards:grafana-dashboard                     grafana_dashboard        regular      
cos-proxy:downstream-prometheus-scrape                    cos-scrape-interval-config-metrics:configurable-scrape-jobs  prometheus_scrape        regular      
cos-proxy:juju-info                                       landscape-client:container                                   juju-info                subordinate  
cos-proxy:juju-info                                       ntp:juju-info                                                juju-info                subordinate  
cos-proxy:juju-info                                       ubuntu-advantage:juju-info                                   juju-info                subordinate  
nrpe:monitors                                             cos-proxy:monitors                                           monitors                 regular      
```

This deployment has
* obs data passed to **cos-proxy**:
  * no dashboard data
* obs data passed by **cos-proxy** elsewhere:
  * **cos-proxy** (provider) *sending metric/log/... settings to* **grafana-agent** (requirer) via the [`cos-agent` interface](https://github.com/canonical/charm-relation-interfaces/tree/main/interfaces/cos_agent/v0) so that grafana-agent can scrape things itself
  * **cos-proxy** sending dashboards that others have passed to it to **cos-grafana-dashboards**

If I understand correctly then, the issue is that **cos-proxy** is configured to be able to send dashboards out to `cos-grafana-dashboards` via the `downstream-grafana-dashboard` relation, but doesn't currently have a source of dashboards, and [according to the charm that should be a Blocked status](https://github.com/canonical/cos-proxy-operator/blob/8174714a6ac85b95b8db2c3ab0c04ae5b4425b14/src/charm.py#L617).  

My first intuition is that this is not a bug exactly, but maybe a design decision we should revisit.  Essentially the charm is raising Blocked if it is properly configured but unused, which feels odd.  I'll check with others though as I might misunderstand this.

FWIW [this pr that landed in March](https://github.com/canonical/cos-proxy-operator/pull/127) changed this logic for rev >=73, and 73 landed into latest/stable about a month ago, so there's a decent chance this is all related.  

If making a change here, we should also revisit the logic for the other required data sources (loki, prometheus) as they probably have the same decisions.

### To Reproduce

Not reproduced, but `juju status` output has detail to reproduce if needed

### Environment

See juju status output

### Relevant log output

```shell
See above
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sed-i**
## Issue
Status is set to Blocked unnecessarily.

## Solution
Use the [mandatory relation pairs helper](https://github.com/canonical/cos-lib/pull/25).

Fixes #144.

## Testing Instructions
See unit tests.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -86,6 +87,32 @@ class COSProxyCharm(CharmBase):
...
+            {"downstream-prometheus-scrape"},
+        ],
+        # "prometheus-rules": [  # must be paired with:
+        #     {"cos-agent"},  # or
+        #     {"downstream-prometheus-scrape"},
+        # ],
+        # "prometheus": [  # must be paired with:
+        #     {"cos-agent"},  # or
+        #     {"downstream-prometheus-scrape"},
+        # ],
```
Comment (sed-i) [inline src/charm.py:113]: Seems like these also have mandatory pairs, but we don't track them with a stored state yet.
Could add this in as a comment or remove entirely. Wdyt?

File: src/charm.py [hunk 2]
```diff
@@ -606,37 +633,41 @@ def _on_nrpe_targets_changed(self, event: Optional[NrpeTargetsChangedEvent]):
...
+            self.unit.status = BlockedStatus(
+                "Missing incoming relations: {}".format(
+                    "|".join(self.mandatory_relation_pairs.keys())
+                )
+            )
+            return
 
-        if (
-            self._stored.have_loki  # pyright: ignore
-            and not (self._stored.have_filebeat or self._stored.have_nrpe)  # pyright: ignore
```
Comment (sed-i) [inline src/charm.py:621]: @IbraAoad @lucabello Iiuc, nrpe doesn't have anything for loki. Am I missing something?

File: src/charm.py (hunk 1 above)
Comment (sed-i) [reply src/charm.py:113]: @simskij could you please review these combos?

File: src/charm.py (hunk 2 above)
Comment (lucabello) [reply src/charm.py:621]: AFAIK you are correct

File: src/charm.py (hunk 1 above)
Comment (lucabello) [reply src/charm.py:113]: LGTM, however I have some doubts which I'll write in a comment below

**Comment by lucabello** [comment]
I think setting Blocked to tell the user their COS Proxy is missing something (i.e., it's "misconfigured" - not as in "broken", but as in "incomplete") is intentional; there are definitely other ways to do that (a message in ActiveStatus?), but I think there's value in telling the user "you haven't fully set up COS Proxy yet"

I don't have a strong opinion at all though, I'd just like to hear from the team before we merge

File: src/charm.py [hunk 3]
```diff
@@ -86,6 +87,32 @@ class COSProxyCharm(CharmBase):
 
     _stored = StoredState()
 
+    mandatory_relation_pairs = {
```
Comment (ca-scribner) [inline src/charm.py:90]: What is the intent of an item in `mandatory_relation_pairs`?  is it:
1. `{"name of mandatory relation": ["additional relations that we require one of", "if we have this relation"]}`
2. {"name of optional relation": ["list of relations that we require one of", "if we have this relation"]}`
?

based on [this](https://github.com/canonical/cos-proxy-operator/pull/145/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR656) I think the intent is definition (1), but I don't think that solves #144.  #144 is a bug where cos-proxy is requiring the user to establish a `dashboards` relation even though they don't have/want one.  

I think here we're combining two concepts:
* a relation that the charm always requires (mandatory)
* a relation that the charm requires iff the charm has some other relation (eg: having one of `cos-agent` or `downstream-grafana-dashboard` if you have a `dashboards` relation)

where iiuc these should be separate.

Comment (ca-scribner) [inline src/charm.py:90]: What is the intent of an item in `mandatory_relation_pairs`?  is it:
1. `{"name of mandatory relation": ["additional relations that we require one of", "if we have this relation"]}`
2. {"name of optional relation": ["list of relations that we require one of", "if we have this relation"]}`
?

based on [this](https://github.com/canonical/cos-proxy-operator/pull/145/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR656) I think the intent is definition (1), but I don't think that solves #144.  #144 is a bug where cos-proxy is requiring the user to establish a `dashboards` relation even though they don't have/want one.  

I think here we're combining two concepts:
* a relation that the charm always requires (mandatory)
* a relation that the charm requires iff the charm has some other relation (eg: having one of `cos-agent` or `downstream-grafana-dashboard` if you have a `dashboards` relation)

where iiuc these should be separate.

**Review by ca-scribner — CHANGES_REQUESTED** [review]

File: src/charm.py [hunk 4]
```diff
@@ -62,6 +62,7 @@
     _type_convert_stored,
 )
 from charms.vector.v0.vector import VectorProvider
+from cosl import MandatoryRelationPairs
```
Comment (PietroPasotti) [inline src/charm.py:65]: I had no idea this existed! TIL :)

File: src/charm.py [hunk 5]
```diff
@@ -606,37 +633,41 @@ def _on_nrpe_targets_changed(self, event: Optional[NrpeTargetsChangedEvent]):
...
+                ("downstream-prometheus-scrape", self._stored.have_prometheus),
+                ("prometheus-target", self._stored.have_targets),
+            ]
+            if indicator
+        }
+
+        if not set(self.mandatory_relation_pairs.keys()).intersection(active_relations):
+            self.unit.status = BlockedStatus(
+                "Missing incoming relations: {}".format(
+                    "|".join(self.mandatory_relation_pairs.keys())
```
Comment (PietroPasotti) [inline src/charm.py:659]:
```suggestion
                    "|".join(self.mandatory_relation_pairs.keys())
```

we should be only listing the difference between what should be and actually is, not the full list, to avoid cluttering.

File: src/charm.py [hunk 6]
```diff
@@ -606,37 +633,41 @@ def _on_nrpe_targets_changed(self, event: Optional[NrpeTargetsChangedEvent]):
             )
 
     def _set_status(self):
```
Comment (PietroPasotti) [inline src/charm.py:635]: why aren't we doing this in collect_unit_status?

File: src/charm.py [hunk 7]
```diff
@@ -606,37 +633,41 @@ def _on_nrpe_targets_changed(self, event: Optional[NrpeTargetsChangedEvent]):
...
-        ):
-            messages.append("one of (Loki|filebeat)")
-
-        if (
-            (self._stored.have_prometheus or self._stored.have_gagent)  # pyright: ignore
-            and not (self._stored.have_targets or self._stored.have_nrpe)  # pyright: ignore
-        ) or (
-            (self._stored.have_targets or self._stored.have_nrpe)  # pyright: ignore
-            and not (self._stored.have_prometheus or self._stored.have_gagent)  # pyright: ignore
+        if missing := MandatoryRelationPairs(self.mandatory_relation_pairs).get_missing_as_str(
```
Comment (PietroPasotti) [inline src/charm.py:664]: is this repeating what's above already? Not sure I see the difference (in semantics).

Comment (sed-i) [reply src/charm.py]: It does not repeat: the first is checking is all incoming relations are missing, and the second is check if any mandatory pairs are missing.

File: src/charm.py [hunk 8]
```diff
@@ -605,38 +666,43 @@ def _on_nrpe_targets_changed(self, event: Optional[NrpeTargetsChangedEvent]):
...
+                ("prometheus", self._stored.have_prometheus_manual),
+            ]
+            if indicator
+        }
+
+        # Set blocked if _all_ incoming relations are missing. This helps notice under-configured
+        # or redundant cos-proxy instances.
+        if not set(self.relation_pairs.keys()).intersection(active_relations):
+            self.unit.status = BlockedStatus("Missing incoming relation(s)")
+            return
```
Comment (Abuelodelanada) [inline src/charm.py:697]: Shouldn't we add a log line in order to be more explicit about how to solve this situation?

**Review by Abuelodelanada — APPROVED** [review]

File: src/charm.py (hunk 8 above)
Comment (sed-i) [reply src/charm.py]: Done.

File: src/charm.py (hunk 3 above)
Comment (ca-scribner) [reply src/charm.py]: That `test_no_incoming_relations_blocks` passes is I think the problem.  If `test_no_incoming_relations_blocks` passes, then #144 will not be fixed

This test describes what #144 needs to pass:
```python
    def test_has_outgoing_dashboard_relation_without_incoming(self):
        """Assert charm is not Blocked if it has outgoing dashboard relation but not incoming."""
        self.harness.set_leader(True)

        # WHEN we have a downstream relation that can process dashboards
        downstream_rel_id = self.harness.add_relation(
            "downstream-grafana-dashboard", "cos-grafana"
        )
        self.harness.add_relation_unit(downstream_rel_id, "cos-grafana/0")

        # AND we have no dashboard relation

        # THEN the charm should not be blocked
        self.harness.evaluate_status()
        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
```

I think the directionality of the mandatory pair is applied backwards here.  The test is saying "if I have a destination for dashboards, then I will block if I don't have a source".  But that's not a misconfigured case.  It would only be misconfigured if we have a source for dashboards and no target to process them

This feels most obvious if we have the `cos-agent` relation.  If we have a `cos-agent` target, then currently we require sources for all of `dashboards`, `monitors`, etc.  That sounds backwards.

Comment (ca-scribner) [reply src/charm.py]: Yeah I see what you mean.  So the more correct test for #144 would be:

```python
    def test_has_outgoing_dashboard_relation_without_incoming(self):
        """Assert charm is not Blocked when we have a valid subset of the supported relations.
         
        This tests that if we have a cos-agent, which can consume a dashboard, but no incoming
        dashboards, that the charm does not block.
        """
        self.harness.set_leader(True)

        # WHEN we have a downstream relation that can process dashboards
        downstream_rel_id = self.harness.add_relation(
            "cos-agent", "cos-agent"
        )
        self.harness.add_relation_unit(downstream_rel_id, "cos-agent/0")

        # AND we have some other producer (a prometheus target)
        rel_id = self.harness.add_relation("prometheus-target", "target-app", app_data={
                "hostname": "scrape_target_0",
                "port": "1234",
            },)

        # AND we have no dashboard relation

        # THEN the charm should not be blocked
        self.harness.evaluate_status()
        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
```

And you're right, it does pass for this PR.

File: src/charm.py [hunk 9]
```diff
@@ -605,38 +666,47 @@ def _on_nrpe_targets_changed(self, event: Optional[NrpeTargetsChangedEvent]):
...
+                ("prometheus-rules", self._stored.have_prometheus_rules),
+                ("prometheus", self._stored.have_prometheus_manual),
+            ]
+            if indicator
+        }
+
+        # Set blocked if _all_ incoming relations are missing. This helps notice under-configured
+        # or redundant cos-proxy instances.
+        if not set(self.relation_pairs.keys()).intersection(active_relations):
+            self.unit.status = BlockedStatus("Missing incoming relation(s)")
```
Comment (ca-scribner) [inline src/charm.py:696]: I found this a little clearer before being updated.  When debugging [this conversation](https://github.com/canonical/cos-proxy-operator/pull/145/files/339cbfbe315a94009b7ad585acb7bfc0279abdf5#r1649170085), I misunderstood this error as the charm requring a specific incoming relation rather than saying I needed one *of any* supported incoming relations. Maybe reword to
```suggestion
            self.unit.status = BlockedStatus("Charm requires at least one incoming relation")
```

or similar?  we could list the relations that satisfy this if we want, or just say it is missing.

File: src/charm.py
Comment (ca-scribner) [inline src/charm.py:1]: Broader discussion based on [this conversation](https://github.com/canonical/cos-proxy-operator/pull/145/files/339cbfbe315a94009b7ad585acb7bfc0279abdf5#r1649170085) - should the charm Block when it has no incoming relations?  

imo the charm is properly configured when it has at least one outgoing consumer relation for every incoming producer, and misconfigured otherwise.  For example:
1. properly configured (Active): 
  a. has an incoming prometheus (`prometheus-target`) and a consumer of prometheus (`downstream-prometheus-scrape`)
  b. has no incoming producers and has a consumer `cos-agent`
  c. has no incoming or outgoing relations
2. misconfigured (Blocked)
  a. has an incoming prometheus and no consumer (no `downstream-prometheus-scrape` or `cos-agent`)

atm 1b and 1c are considered misconfigured, but imo they should be correctly configured.  I compare it to prometheus the workload - if I deploy prometheus but do not yet have any data going to it, prometheus's health checks still report the workload is running fine

Comment (ca-scribner) [reply src/charm.py:1]: Summary of an offline followup:
* grafana-agent has a similar design where the absense of incoming relations is considered misconfiguration, so that is replicated here

**Review by ca-scribner — APPROVED** [review]
lgtm.  there's a minor suggestion outstanding about error language, but its not a blocker

File: src/charm.py (hunk 9 above)
Comment (sed-i) [reply src/charm.py:696]: Done in next commit.

File: src/charm.py (hunk 3 above)
Comment (sed-i) [reply src/charm.py]: Test added in next commit.
